crates/hyperdrive-math/src/lp.rs

```diff
@@ -600,6 +621,567 @@ impl State {
 
         Ok(derivative)
     }
+
+    /// Checks to see if we should short-circuit the iterative calculation of
+    /// the share proceeds when distributing excess idle liquidity. This
+    /// verifies that the ending LP share price is greater than or equal to the
+    /// starting LP share price and less than or equal to the starting LP share
+    /// price plus the minimum tolerance.
+    fn should_short_circuit_distribute_excess_idle_share_proceeds(
+        &self,
+        active_lp_total_supply: FixedPoint, // just the total number of lp shares
+        starting_present_value: FixedPoint,
+        lp_total_supply: FixedPoint, // total lp shares and withdrawal shares - w.s. ready to withraw
+        present_value: FixedPoint,
+    ) -> Result<bool> {
+        if active_lp_total_supply == fixed!(0) {
+            return Err(eyre!("active_lp_total_supply is zero"));
+        }
+        if lp_total_supply == fixed!(0) {
+            return Err(eyre!("lp_total_supply is zero"));
+        }
+        let new_lp_share_price_down = present_value.div_down(active_lp_total_supply);
+        let new_lp_share_price_up = present_value.div_up(active_lp_total_supply);
+        let old_lp_share_price = starting_present_value.div_up(lp_total_supply);
+        let old_lp_share_price_with_tolerance = (fixed!(1e18)
+            + FixedPoint::from(SHARE_PROCEEDS_SHORT_CIRCUIT_TOLERANCE))
+        .mul_div_down(starting_present_value, lp_total_supply);
+
+        Ok(
+            // Ensure that new LP share price is greater than or equal to the
+            // previous LP share price:
+            //
+            // PV_1 / l_1 >= PV_0 / l_0
+            //
+            // NOTE: Round the LHS down to make the check stricter.
+            new_lp_share_price_down >= old_lp_share_price
+            // Ensure that new LP share price is less than or equal to the
+            // previous LP share price plus the minimum tolerance:
+            //
+            // PV_1 / l_1 <= (PV_0 / l_0) * (1 + tolerance)
+            //
+            // NOTE: Round the LHS up to make the check stricter.
+            && new_lp_share_price_up <= old_lp_share_price_with_tolerance,
+        )
```

Comment:
It would be better if this more closely matched the Solidity IMO. Re-ordering the logic makes this quite a bit harder to review.